sqldb/sqlerrors.go

```diff
@@ -0,0 +1,71 @@
+package sqldb
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgerrcode"
+	"modernc.org/sqlite"
+	sqlite3 "modernc.org/sqlite/lib"
+)
+
+// MapSQLError attempts to interpret a given error as a database agnostic SQL
+// error.
+func MapSQLError(err error) error {
+	// Attempt to interpret the error as a sqlite error.
+	var sqliteErr *sqlite.Error
+	if errors.As(err, &sqliteErr) {
+		return parseSqliteError(sqliteErr)
+	}
+
+	// Attempt to interpret the error as a postgres error.
+	var pqErr *pgconn.PgError
+	if errors.As(err, &pqErr) {
+		return parsePostgresError(pqErr)
+	}
+
+	// Return original error if it could not be classified as a database
+	// specific error.
+	return err
+}
+
+// parsePostgresError attempts to parse a sqlite error as a database agnostic
+// SQL error.
+func parseSqliteError(sqliteErr *sqlite.Error) error {
+	switch sqliteErr.Code() {
+	// Handle unique constraint violation error.
+	case sqlite3.SQLITE_CONSTRAINT_UNIQUE:
```

Comment:
I wonder if there's added value from mapping the unique constraint error at all? Apart from logging of course.

Reply:
IIRC, we ended up doing this in the tap repo so we could get better logging from the DB, otherwise it's some like single integer return error from `sqlite`.